File: starters/Java/bomberjam/Game.java
package bomberjam;

import com.google.gson.Gson;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Game {
    private static final Gson GSON = new Gson();

    private final BufferedReader stdin;
    private boolean isReady;
    private String myPlayerId;
    private State state;

    public Game() {
        this.stdin = new BufferedReader(new InputStreamReader(System.in));
        this.isReady = false;
        this.myPlayerId = null;
        this.state = null;
    }

    public String getMyPlayerId() {
        this.ensureIsReady();
        return this.myPlayerId;
    }

    public State getState() {
        this.ensureIsReady();
        this.ensureInitialState();
        return this.state;
    }

    public Player getMyPlayer() {
        this.ensureIsReady();
        this.ensureInitialState();
        return this.state.getPlayers().get(this.myPlayerId);
    }

    public void setReady(String playerName) throws Exception {
        if (this.isReady)
            return;

        if (playerName == null || playerName.length() == 0)
            throw new IllegalArgumentException("Your name cannot be null or empty");

        System.out.println("0:" + playerName);
        System.out.flush();

        this.myPlayerId = stdin.readLine();

        if (!isInteger(this.myPlayerId))
            throw new IllegalStateException("Could not retrieve your ID from standard input");

        this.isReady = true;
    }

    private static boolean isInteger(String text) {
        if (text == null) {
            return false;
        }

        try {
            int i = Integer.parseInt(text);
        } catch (NumberFormatException nfe) {
            return false;
        }

        return true;
    }

    public void receiveCurrentState() throws Exception
    {
        this.ensureIsReady();
        this.state = GSON.fromJson(this.stdin.readLine(), State.class);
    }

    public void sendAction(ActionKind action)
    {
        this.ensureIsReady();
        this.ensureInitialState();

        String actionStr = Constants.ACTION_KIND_TO_ACTION_STRING_MAPPINGS.get(action);

        System.out.println(this.state.getTick() + ":" + actionStr);
        System.out.flush();
    }

    private void ensureIsReady()
    {
        if (!this.isReady)
            throw new IllegalStateException("You need to call Game.ready(...) with your name first");
    }

    private void ensureInitialState()
    {
        if (this.state == null)
            throw new IllegalStateException("You need to call Game.receiveCurrentState() to retrieve the initial state of the game");
    }
}

File: starters/JavaScript/bomberjam/game.js
const rl = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout,
});

const readlineAsync = (function () {
  const readlineGen = (async function* () {
    for await (const line of rl) {
      yield line;
    }
  })();
  return async () => ((await readlineGen.next()).value);
})();

class Game {
  constructor() {
    this._isReady = false;
    this._myPlayerId = null;

    /** @type {State} */
    this._state = null;
  }

  /** @returns {string} */
  get myPlayerId() {
    this._ensureIsReady();
    return this._myPlayerId;
  }

  /** @returns {State} */
  get state() {
    this._ensureIsReady();
    this._ensureInitialState();
    return this._state;
  }

  /** @returns {Player} */
  get myPlayer() {
    this._ensureIsReady();
    this._ensureInitialState();
    return this._state.players[this._myPlayerId];
  }

  /**
   * @param {string} playerName
   * @returns {Promise<void>}
   */
  async ready(playerName) {
    if (this._isReady)
      return;

    if (typeof playerName !== 'string' || playerName.trim().length === 0)
      throw new Error('Your name cannot be null or empty');

    console.log('0:' + playerName);

    this._myPlayerId = await readlineAsync();

    if (!/^\d+$/.test(this._myPlayerId))
      throw new Error('Could not retrieve your ID from standard input');

    this._isReady = true;
  }

  /** @returns {Promise<void>} */
  async receiveCurrentState() {
    this._ensureIsReady();
    const jsonState = await readlineAsync();
    this._state = JSON.parse(jsonState);
  }

  /** @param {string} action */
  sendAction(action) {
    this._ensureIsReady();
    this._ensureInitialState();

    console.log(this._state.tick + ':' + action);
  }

  _ensureIsReady() {
    if (!this._isReady)
      throw new Error('You need to call Game.ready(...) with your name first');
  }

  _ensureInitialState() {
    if (!this._state)
      throw new Error('You need to call Game.receiveCurrentState() to retrieve the initial state of the game');
  }

  close() {
    rl.close();
  }
}

module.exports = Game;